keps/sig-apps/4212-improve-node-maintenance/README.md

```diff
@@ -0,0 +1,1065 @@
+
+
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a cluster admin I want to have a simple interface to initiate a node drain/maintenance without
+any required manual interventions. I also want to observe the node drain via the API and check on
+its progress. I also want to be able to discover workloads that are blocking the node drain.
+
+#### Story 2
+
+As an application owner, I want to run single replica applications without disruptions and have the
+ability to easily migrate the workload pods from one node to another.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+### Kubectl
+
+`kubectl drain` command will be changed to create a NodeMaintenance object after marking node
+unschedulable. We will also change the implementation to skip applications that support
+workload migration. At first, we will just check deployments with a positive
+`.spec.strategy.rollingUpdate.maxSurge` value. To make this extensible, we propose a new Surge API
+that could be checked before disrupting any workload (see below). If the cluster,
+doesn't support the NodeMaintenance API, kubectl will perform the node drain in a backwards
+compatible way.
+
+`kubectl uncordon` will be amended with `--node-maintenance` option. This can be used to delete the
+NodeMaintenance object together with marking the node as schedulable again. The default value of
+this flag will be `false`.
+
+### NodeMaintenance API
+
+NodeMaintenance objects serve as an intent to remove or migrate pods from a particular node. We will
+also support setting a reason for the maintenance. It is unclear whether we should allow marking the
+maintenances as finished and keeping the old maintenances in the system for bookkeeping purposes.
+
+As an alternative, it would be possible to mark the node object instead of this API. But it is
+probably better to decouple this from the node for extensibility reasons. As we can see, the
+`kubectl drain` logic is a bit complex, and it may be possible to move this logic to a controller
+in the future and make the node maintenance purely declarative. It also helps to decouple RBAC
+permissions from the node object.
+
+```go
+
+type NodeMaintenance struct {
+    ...
+    Spec NodeMaintenanceSpec
+    Status NodeMaintenanceStatus
+}
+
+type NodeMaintenanceSpec struct {
+    NodeName string
+	Reason string
```

Comment:
If we want to support a node selector here, is there a way to add that support?

Reply:
I am not really sure if we want to support the node selector. It would make everything more complex. I think it is better to leave the node selection logic to the node orchestrator, and it should not be a big deal for it to create multiple NodeMaintenance objects.

It might be possible to add the node selector support later if we wish, but hard to do it the other way around.

Still, I would like to hear other people's opinions on this as well.

Reply:
What I mean is a [for but not with](https://en.wikipedia.org/wiki/For_but_not_with) design: making sure that even if this API reaches v1 / GA, there **is** a feasible way to have a node selector.

So long as the API leaves room for it, we do not need to implement the selector logic or API field at all.

Reply:
Got it. This is a very good remark - we should make the API general enough so that the support can be added later. And let's keep this information in mind at every step (API, implementation, documenation) so we don't unknowingly put up barriers to it. 

I have updated the ` NodeMaintenance API` section in the KEP to reflect this.

Reply:
Thanks. I consider this resolved.